project-ideas/order-ledger/createSalesOrder.md

```diff
@@ -1,47 +1,210 @@
-3. Set following to orderContext
-    * orderContext.externalId = orderJson.externalId
-    * orderContext.orderTypeId = "SALES_ORDER"
-    * orderContext.orderName = orderJson.name
-    * orderContext.salesChannelEnumId = orderJson.salesChannelEnumId
-    * orderContext.orderDate = orderJson.createdAt
-    * orderContext.currencyUom = orderJson.currencyUom
-    * orderContext.presentmentCurrencyUom = orderJson.presentmentCurrencyUom
-    * orderContext.grandTotal = orderJson.grandTotal
-    * orderContext.remainingSubTotal = orderJson.remainingSubTotal
-    * orderContext.statusId = orderJson.statusId
-    * orderContext.productStoreId = orderJson.productStoreId
-    * orderContext.attributes = orderJson.attributes
-    * orderContext.adjustments = orderJson.adjustments
-4. Initialize orderContext.roles list
-    * Call findOrCreate#Person for orderJson.customer
-    * Add [partyId:findOrCreatePartyOutput.partyId, roleTypeId:"SHIP_TO_CUSTOMER"] to orderContext.roles
+        * payload (Map)
+2. Validate order by checking OrderHeader.externalId to avoid duplication of order
+3. Initialize payload.roles list
+    * Call findOrCreate#Person for payload.customer
+    * Add [partyId:findOrCreatePersonOutput.partyId, roleTypeId:"SHIP_TO_CUSTOMER"] to payload.roles
     * Get ProductStore for orderJson.productStoreId
-    * Add [partyId:ProductStore.payToPartyId, roleTypeId:"SHIP_FROM_VENDOR"] to orderContext.roles
-5. Initialize orderContext.contactMechs list
-    * For orderJson.shipToAddress map call create#PostalAddress with input as orderJson.shipToAddress
-    * Set shipToAddressContactMechId = createPostalAddressOutput.contactMechId
-    * Add [contatctMechId:shipToAddressContactMechId, contactMechPurposeTypeId:"SHIPPING_LOCATION"] to orderContext.contactMechs
-    * Call create#ContactMech for [contactMechTypeId:"TELECOM_NUMBER", infoString:orderJson.shipToPhone]
-    * Set shipToPhoneContactMechId = createContactMechOutput.contactMechId
-    * Add [contatctMechId:shipToPhoneContactMechId, contactMechPurposeTypeId:"PHONE_SHIPPING"] to orderContext.contactMechs
-    * Call create#ContactMech for [contactMechTypeId:"EMAIL_ADDRESS", infoString:orderJson.email]
-    * Add [contatctMechId:createContactMechOutput.contactMechId, contactMechPurposeTypeId:"ORDER_EMAIL"] to orderContext.contactMechs
-    * If orderJson.billToAddress
-        * For orderJson.billToAddress map call create#PostalAddress with input as orderJson.billToAddress
-        * Add [contatctMechId:createPostalAddressOutput.contactMechId, contactMechPurposeTypeId:"BILLING_LOCATION"] to orderContext.contactMechs
-    * If orderJson.billToPhone
-        * Call create#ContactMech for [contactMechTypeId:"TELECOM_NUMBER", infoString:orderJson.billToPhone]
-        * Add [contatctMechId:createContactMechOutput.contactMechId, contactMechPurposeTypeId:"PHONE_BILLING"] to orderContext.contactMechs
+    * Add [partyId:ProductStore.payToPartyId, roleTypeId:"SHIP_FROM_VENDOR"] to payload.roles
```

Comment:
If we have customer Party in our database, we may already have all these details. 

lookup can be super expensive. 

Is it better to just create on every next order by same person? 

Or is there efficient algorithm to lookup / compare? 

What if we create hash of the elements in contact mach and  then compare the hash. We did similar thing in Product sync?

Reply:
We are creating contact mechs for the order and not for customer. Do you mean we dont need to create the same address eg, shipping address in every order, and check if any way to re-use the same address id and just associate the order and the contact mech id?

Reply:
The OMS only stores customer information in relation to a single transaction. The only customer related detail re-used across orders is party id. This way, if we ever need to purge a customer from our system all we need to do is sanitize their party and person record without tampering with order related details like ship to address